src/app/SOCestimator/SOCestimator.c

```diff
@@ -0,0 +1,134 @@
+    }
+    else if (avg_V < SOC_75_LIMIT_FOR_TEMP_INDEX(temp_index))
+    {
+        return 75;
+    }
+    else
+    {
+        return 100;
+    }
+}
+
+/**
+ * Calculate the change in SOC (delta SOC) resulting from a current held over an elapsed time.
+ * current_A [in] - current held
+ * time_elapsed_ms [in] - time elapsed
+ */
+static inline float calculate_dSOC(float current_A, unsigned int time_elapsed_ms)
+{
+    return (current_A * time_elapsed_ms * 0.001) / BATTERY_CAPACITY_Ah;
+}
+
+// SOC waits for an average voltage reading to determine starting SOC.
+// This flag stays false until this first average voltage reading is recieved.
+// Current readings up to this point are aggregated and applied to the starting SOC model
+// as soon as the first voltage is received.
+bool starting_soc_calculated = false;
+// keep track of current and time expired prior to the initial SOC estimate
+float pre_init_current_A = 0.0;
+
+void SOCestimator_init(void)
+{
+    SOC_raw = 100;
+    SOC_limit = 100;
+    starting_soc_calculated = false;
+}
+
+/**
+ * Use the instantaneous current to update the coulomb counting SOC model.
+ * current [in] - the instantaneous current (negative for discharge, positive for charging)
+ */
+void SOCestimator_coulomb_count_update_1kHz(float current_A)
+{
+    if (starting_soc_calculated)
+    {
+        SOC_raw = SOC_raw + calculate_dSOC(current_A, 1);
+    }
+    else
+    {
+        pre_init_current_A += current_A;
+    }
+}
+
+/**
+ * Calculate the average voltage of the battery cells. Limit the SOC measurement to
+ * the highest SOC possible at this voltage.
+ * battery_model [in] - new cell voltages to use in calculating the SOC
+ * ambient_temp [in] - ambient temperature
+ */
+void SOCestimator_voltage_threshold_update_10Hz(BatteryModel_t* battery_model, float ambient_temp_C)
+{
+    // calculate the average series cell voltage
+    float total_V = 0.0;
+    for (int i = 0; i < NUM_SERIES_CELLS; i++)
+    {
+        total_V += battery_model->cells[i].voltage;
+    }
+
+    float avg_V = total_V / (float) NUM_SERIES_CELLS;
+
+    int new_soc_limit = SOC_limit_from_voltage(avg_V, ambient_temp_C);
+
+    // limit may flicker here
+    SOC_limit = new_soc_limit;
+
+    if (!starting_soc_calculated)
+    {
+        // This is the first voltage calculated.
+        // Change the SOC as if all the pre-init current readings happened over 1ms.
+        // This only works because the current measurements are all over 1ms.
+        SOC_raw = calculate_dSOC(pre_init_current_A, 1);
+
+        starting_soc_calculated = true;
+    }
+}
+
+/**
+ * Get the SOC as calculated by coulomb counting without voltage threshold limit.
+ * return int [0:100]
+ */
+int SOCestimator_get_soc_raw(void)
+{
+    return SOC_raw;
+}
+
+/**
+ * Get the SOC as calculated by coulomb counting with the voltage threshold limit applied.
+ * return int [0:100]
+ */
+int SOCestimator_get_soc_corrected(void)
+{
+    return MIN(SOC_limit, SOC_raw);
```

Comment:
So is this SOC_limit to account for the effect of temperature on the SOC? If so it seems weird since SOC_limit can only be 25, 50, or 75.

Reply:
it's a sanity check on the coulomb counting based on the average cell voltage. The average cell voltage v.s. SOC chart on the data sheet varies based on temperature